design/sticky-session-design.md

```diff
@@ -0,0 +1,343 @@
+# Sticky session support
+
+**Status**: _draft_
+
+This document proposes adding support for sticky session in contour.
+
+## Goals
+
+- Enabling support for sticky-sessions in Contour using
+    - Cookie
+    - Source IP
+    - Header
+- Adding support for sticky-sessions in IngressRoute CRD. 
+
+## Non-goals
+
+- Making options in IngressRoute very specific to Envoy.
+
+## Background
+
+Sticky-sessions allow requests to be routed to the same host based on certain parameters of the request.
+The parameters can be a cookie, the source IP, a custom header, etc.
+
+#### Sticky sessions in Envoy 
+Envoy supports sticky sessions by using a hash-based load balancing strategy (RingHash, Maglev). 
+Specifying a **hash policy** for such load-balancers will enable sticky sessions. Hash policy 
+is specified per route. **For sticky sessions, each route should only have one service
+associated to the route**. This is because if multiple clusters are associated per route (weighted clusters),
+requests will be forwarded to the cluster based on weights and stickiness will be 
+enforced in that cluster from that point. This defeats the purpose of using sticky sessions since a session will not
+be pinned to a pod. 
+
+## High-Level Design
+In contour, sticky sessions will be enabled per service. 
+If a service has sticky sessions enabled, it should be the **only service**
+under that route. Sticky sessions **do not** work with multiple services (weighted clusters).
+This means that sticky sessions cannot be used along with blue/green deployments.
+
+### Sticky-sessions configuration
+To enable sticky-sessions in a service, the `strategy` needs to be set to *StickySession*.
+By default, sticky-session will be based on **source-ip**. `stickySessionConfig` can 
```

Comment:
I understand your point. If we use cookie by default, can you suggest the default cookie name to use?

Reply:
>  If we use cookie by default, can you suggest the default cookie name to use?

This will have to be supplied by the customer, it's their cookie we're piggybacking on top of.

Reply:
I meant that if in case a user does not provide any configuration associated with the sticky-session, what should be the default behaviour? Should we set an error or use a default parameter to hash on?
Note that the user will need to mention the strategy as _StickySession_ and additionally provide _stickySessionConfig_.